src/values/ArrayValue.js

```diff
@@ -44,8 +50,16 @@ function evaluatePossibleNestedOptimizedFunctionsAndStoreEffects(
     }
     invariant(funcToModel instanceof ECMAScriptSourceFunctionValue);
 
-    if (realm.instantRender.enabled && realm.collectedNestedOptimizedFunctionEffects.has(funcToModel)) {
-      realm.instantRenderBailout("Array operators may only be optimized once", funcToModel.expressionLocation);
+    if (realm.collectedNestedOptimizedFunctionEffects.has(funcToModel)) {
+      if (realm.instantRender.enabled) {
+        realm.instantRenderBailout("Array operators may only be optimized once", funcToModel.expressionLocation);
+      } else {
+        realm.collectedNestedOptimizedFunctionEffects.delete(funcToModel);
```

Comment:
So you uncollect it here so that it doesn't get optimized with a captured state.
But what if the function gets treated like this for a third time?